controller: DEV_SET_CONFIG: remove old configs for AL-MAC

CoralMalachi opened the pull request:
currently, test 5.4.2 keeps on failing with the following message :

![image](https://user-images.githubusercontent.com/38204874/68112282-aaf4ea00-fef9-11e9-99b1-8b5a464aa524.png)


It means that the SSID was not changed as it was supposed to.
The solution is to remove all old configs for the AL-MAC address when we
get a DEV_SET_CONFIG command.

After this change, it was verified that the AP configuration renew
succeeded and the SSID was changed as required.
Fixes #195 

In addition, this PR is also fixed #199, by adding a validation check for the 
bss_info_conf content.


Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
controller: db: add clear bss info by almac address method

Preparative commit.
Add a method that clears all old configs for a specific AL-MAC address.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

arnout commented:
I haven't reviewed the details yet, but one overall comment: instead of prepending to the list, we should remove all old configs for the AL-MAC address when we get a DEV_SET_CONFIG command.

tomereli reviewed (changes requested):
please update PR headline to reflect the new changes we do here, fix issues with CI and test it at least once in the certification testbed before we actually merge it.

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -3190,6 +3190,8 @@ std::list<db::bss_info_conf_t> &db::get_bss_info_configuration(const sMacAddr &a
 
 void db::clear_bss_info_configuration() { bss_infos.clear(); }
 
+void db::clear_bss_info_configuration(const sMacAddr &al_mac) { bss_infos[al_mac].clear(); }
```
It will be more suitable to edit the exist clear function and add it to it an optional value of MacAddr.
Of course that it will cause the function to be with more that one line so I will be needed to move the implementation to the CPP file.

arnout replied:
> It will be more suitable to edit the exist clear function and add it to it an optional value of MacAddr.

I disagree, I think it makes more sense that it's an overload. An optional argument has two disadvantages:

- It needs to be a pointer (otherwise it's not optional) which makes the code slightly more difficult to read.
- The implementation will simply have a condition on the argument being nullptr or not; two different function signature are much easier to read.

> Of course that it will cause the function to be with more that one line so I will be needed to move the implementation to the CPP file.

It already is in the cpp file :-)

morantr replied:
> It already is in the cpp file :-)

Yeah I confused 😏 

> >It will be more suitable to edit the exist clear function and add it to it an optional value of MacAddr.
> 
> I disagree, I think it makes more sense that it's an overload. An optional argument has two disadvantages:
> 
> It needs to be a pointer (otherwise it's not optional) which makes the code slightly more difficult to read.
> The implementation will simply have a condition on the argument being nullptr or not; two different function signature are much easier to read.

It does not have to be a pointer, it can still be a reference. I tested the following code and it worked:

```cpp
#include <iostream>
#include <string>

std::string ret_str(const std::string &str = "hello world") 
{
  return str;
}

int main() {
  std::cout << ret_str() << std::endl;
}
```
`output: hello world`

arnout replied:
> > > It will be more suitable to edit the exist clear function and add it to it an optional value of MacAddr.
> > I disagree, I think it makes more sense that it's an overload. An optional argument has two disadvantages:
> > It needs to be a pointer (otherwise it's not optional) which makes the code slightly more difficult to read.
> It does not have to be a pointer, it can still be a reference. I tested the following code and it worked:
> ```c++
> std::string ret_str(const std::string &str = "hello world") 
> ```

Yes, but what would you give as default?

```
    void clear_bss_info_configuration(const sMacAddr &al_mac = ????);
```

You could use `ZERO_MAC`, indeed, then the issue with the pointer is gone. But still you would need a condition in the implementation, and you'd need it as well in the documentation (which is now missing anyway). If you have it as separate function signatures, the condition is implied.

Clearly this is a nitpicking issue, but it would be good to decide if we in general prefer optional arguments to be declared with default parameter values and a condition in the implementation, or as two separate overloaded functions (where one potentially calls the other).

morantr commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -708,7 +710,11 @@ void wfa_ca::handle_wfa_ca_message(
+            if (!old_configs_removed) {
+                database.clear_bss_info_configuration(net::network_utils::mac_from_string(al_mac));
+                old_configs_removed = true;
+            }
```
What is the purpose of the flag `old_configs_removed` and the `if` statement? 
`old_configs_removed` is always set to `false` a couple of lines before, so the code inside the if statement will always happen, so it makes the `if` statement redundant. Also in the `if` scope you change the value old_configs_removed  to `true` but it is not used afterward so it turns out that this flag is redundant too.

CoralMalachi replied:
I did it since I want to remove the old configuration and then set the new one.
In each DEV_SET_CONFIG there are two SSID one for each VAP,
If I call the clear method in each iteration, at the second iteration I'll remove the first SSID I just added to the database, so the boolean variable is used to clear the old configuration the right way.
If you have a better way to do this I'll try it .

morantr replied:
Oh sorry, you are right. I didn't notice that the `old_configs_removed` is defined outside the loop.
But still, this solution has a hole in it.
Since you used a flag, which you set it one time, it will create a problem if there are more than one AL_MACs. So, in that case, it will work for the first AL_MAC, but for the second, third and go on, it will not delete the configuration.
I suggest creating unordered_set of al_mac which we have cleared configuration on:

```cpp
std::unordered_set<std::string> al_mac_cleared_conf;
// code 
// for loop
    // code
    if (al_mac_cleared_conf.find(al_mac) == al_mac_cleared_conf.end()) {
          database.clear_bss_info_configuration(net::network_utils::mac_from_string(al_mac));
          al_mac_cleared_conf.insert(al_mac);
    }
```

morantr reviewed (changes requested):
I don't understand one thing. Before the UCC set config it should remove the config first with DEV_RESET_DEFAULT, so how come this bug happens?

CoralMalachi commented:
> I don't understand one thing. Before the UCC set config it should remove the config first with DEV_RESET_DEFAULT, so how come this bug happens?

The UCC does it only at the beginning of the test before the first DEV_SET_CONFIG.
In 5.4.2 test the UCC sends DEV_SET_CONFIG twice, and before the second time there is no DEV_RESET_DEFAULT, so the old config is still there...

Commit:
controller: validate bss info

Currently, test 5.4.3 fails with the following message :

Beacon frame found, check CTT Agent1 is not operating on the configured channel

The UCC tries to validate there are no beacons from that radio, but
beacon frame found.

The reason for this is when the UCC sends the command dev_set_config
without SSID argument it means we should tear down the specified radio,
but the current implementation does not support it.

The solution is to tear down the radio when the UCC sends the
DEV_SET_CONFIG without the SSID parameter, instead of setting the SSID
to an empty string.

Therefore, add a validation check for the bss_info_conf content.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

tomereli reviewed (approved):
Fixes #195 - that's not entirely true. Say that it fixes some of the issues which are causing #195 to fail.
Other than that - looks fine.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -621,7 +621,7 @@ bool master_thread::autoconfig_wsc_add_m2(std::shared_ptr<ieee1905_1::tlvWscM1>
     uint8_t buf[1024];
     WSC::cConfigData config_data(buf, sizeof(buf), false, true);
-    if (bss_info_conf) {
+    if (bss_info_conf && !bss_info_conf->ssid.empty() && !bss_info_conf->network_key.empty()) {
```
I don't like this. I think this empty SSID is an artifact of the CAPI and not generic enough to warrant this special treatment here. So I think the empty SSID should be handled in handle_dev_set_config() instead.

Also, an empty network key definitely doesn't mean that it should be cleared. If it's an open network, there will be no network key.

arnout reviewed (changes requested):
Note that this needs to be redone entirely because the code has moved. So please rebase on master before resubmitting for review.

I think the commit messages could be much improved by saying what commands UCC sends and what result those commands are expected to have. The details of how this is found in the certification logs are not so important. So, for example, for the first fix, it could be:

```
UCC sends the following commands:

dev_set_config,program,map,bss_info1,be:66:8b:4c:f7:0e 8x Multi-AP-24G-1 0x0020 0x0008 maprocks1 0 1,bss_info2,be:66:8b:4c:f7:0e 8x Multi-AP-24G-2 0x0020 0x0008 maprocks2 1 0
...
dev_set_config,program,map,bss_info1,be:66:8b:4c:f7:0e 8x Multi-AP-24G-1-diff 0x0020 0x0008 maprocks1 0 1,bss_info2,be:66:8b:4c:f7:0e 8x Multi-AP-24G-2-diff 0x0020 0x0008 maprocks2 1 0

The intention is that the second `dev_set_config` replaces 
the first one entirely, i.e. all bss_info for the specified AL-MAC are removed.
However, currently we append the new bss_info to the existing ones.
Therefore, clear the existing bss_info when dev_set_config is done.
```

Similar for the second fix.

Also I wonder if it wouldn't be possible to simply clear the entire bss_info for all AL-MAC when the dev_set_config is given. But it's still not clear to me what the intention is for UCC. I think the only case where dev_set_config is given multiple times is in 5.4.2 and 5.4.3, where the intention really is to clear it.

Commit:
controller: DEV_SET_CONFIG: remove old configs for AL-MAC

UCC sends the following commands:

dev_set_config,program,map,bss_info1,be:66:8b:4c:f7:0e 8x Multi-AP-24G-1 0x0020 0x0008 maprocks1 0 1,bss_info2,be:66:8b:4c:f7:0e 8x Multi-AP-24G-2 0x0020 0x0008 maprocks2 1 0
...
dev_set_config,program,map,bss_info1,be:66:8b:4c:f7:0e 8x Multi-AP-24G-1-diff 0x0020 0x0008 maprocks1 0 1,bss_info2,be:66:8b:4c:f7:0e 8x Multi-AP-24G-2-diff 0x0020 0x0008 maprocks2 1 0

The intention is that the second `dev_set_config` replaces
the first one entirely, i.e. all bss_info for the specified AL-MAC are removed.
However, currently we append the new bss_info to the existing ones.
Therefore, clear the existing bss_info when dev_set_config is done.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
tools: docker: add tear down flow test

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

arnout commented on controller/src/beerocks/master/controller_ucc_listener.cpp:
```diff
@@ -80,6 +81,7 @@ bool controller_ucc_listener::send_cmdu_to_destination(Socket *sd,
 bool controller_ucc_listener::handle_dev_set_config(
     std::unordered_map<std::string, std::string> &params, std::string &err_string)
 {
+    std::unordered_set<std::string> al_mac_cleared_conf;
```
I would have used sMacAddr instead of std::string, but that's probably nitpicking.

Hang on, it's not nitpicking at all: you need the first commit because you're using a string here instead of sMacAddr! This shows that there is a good reason to keep both the upper and lower case in the test.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -630,15 +630,15 @@ bool master_thread::autoconfig_wsc_add_m2(std::shared_ptr<ieee1905_1::tlvWscM1>
 
         LOG(DEBUG) << "WSC config_data:" << std::hex << std::endl
-                   << "     ssid: " << config_data.ssid() << std::endl
+                   << "     ssid: " << config_data.ssid_str() << std::endl
```
I would have done this change in a separate commit, but again that's nitpicking.

arnout commented on tools/docker/tests/test_flows.sh:
```diff
@@ -59,7 +59,7 @@ test_initial_ap_config() {
     # Configure the controller and send renew
     send_CAPI_command gateway "DEV_RESET_DEFAULT" $redirect
-    send_CAPI_command gateway "DEV_SET_CONFIG,bss_info1,$MAC_AGENT1 8x Multi-AP-24G-1 0x0020 0x0008 maprocks1 0 1,bss_info2,$mac_agent1 8x Multi-AP-24G-2 0x0020 0x0008 maprocks2 1 0" $redirect
+    send_CAPI_command gateway "DEV_SET_CONFIG,bss_info1,$MAC_AGENT1 8x Multi-AP-24G-1 0x0020 0x0008 maprocks1 0 1,bss_info2,$MAC_AGENT1 8x Multi-AP-24G-2 0x0020 0x0008 maprocks2 1 0" $redirect
```
NACK this. The whole point of this MAC_AGENT1 is to verify that it is really case insensitive, so we want *both* upper and lower case version in the same configuration.

arnout commented on tools/docker/tests/test_flows.sh:
```diff
@@ -82,8 +82,34 @@ test_ap_config_renew() {
-    err "ap_config_bss_tear_down not implemented yet."
-    return 0
+    status "test ap config bss tear down"
+    # Regression test: MAC address should be case insensitive
```
To make it a real regression test, I think it's better to add it with one case and to remove it with a different case. So the simplest would be to only use the uppercase version in the second dev_set_config.

arnout commented on tools/docker/tests/test_flows.sh:
```diff
@@ -82,8 +82,34 @@ test_ap_config_renew() {
+    MAC_AGENT1=$(echo $mac_agent1 | tr a-z A-Z)
+    # Configure the controller and send renew
+    send_CAPI_command gateway "DEV_RESET_DEFAULT" $redirect
+    send_CAPI_command gateway "DEV_SET_CONFIG,bss_info1,$MAC_AGENT1 8x Multi-AP-24G-1 0x0020 0x0008 maprocks1 0 1,bss_info2,$MAC_AGENT1 8x Multi-AP-24G-2 0x0020 0x0008 maprocks2 1 0" $redirect
```
```suggestion
    send_CAPI_command gateway "DEV_SET_CONFIG,bss_info1,$mac_agent1 8x Multi-AP-24G-1 0x0020 0x0008 maprocks1 0 1,bss_info2,$mac_agent1 8x Multi-AP-24G-2 0x0020 0x0008 maprocks2 1 0" $redirect
```

And maybe a single BSS is sufficient too, so remove bss_info2.

arnout commented on tools/docker/tests/test_flows.sh:
```diff
@@ -82,8 +82,34 @@ test_ap_config_renew() {
+    # SSIDs have been removed for the CTT Agent1's front radio
+    send_CAPI_command gateway "DEV_SET_CONFIG,bss_info1,$MAC_AGENT1 8x" $redirect
+    # Send renew message
+    send_CAPI_command gateway "DEV_SEND_1905,DestALid,$MAC_AGENT1,MessageTypeValue,0x000A,tlv_type1,0x01,tlv_length1,0x0006,tlv_value1,0x${gw_mac_without_colons},tlv_type2,0x0F,tlv_length2,0x0001,tlv_value2,{0x00},tlv_type3,0x10,tlv_length3,0x0001,tlv_value3,{0x00}}" $redirect
```
```suggestion
    send_CAPI_command gateway "DEV_SEND_1905,DestALid,$mac_agent1,MessageTypeValue,0x000A,tlv_type1,0x01,tlv_length1,0x0006,tlv_value1,0x${gw_mac_without_colons},tlv_type2,0x0F,tlv_length2,0x0001,tlv_value2,{0x00},tlv_type3,0x10,tlv_length3,0x0001,tlv_value3,{0x00}}" $redirect
```

arnout reviewed (approved):
It's an excellent idea to add a test for the teardown flow!

Even though I have change requests that are essential, I'm going to approve now already because the fixes are trivial to implement, and because I won't be back until Tuesday to approve an updated version. @morantr can do the final review.
